### ISSUE TITLE
Poor accuracy of pointer X/Y location inference

### ISSUE THREAD

**Issue by ahsin-s** [OthersideAI/self-operating-computer#102]
the X/Y coordinates inferred by the model are always off. It can't even select the address bar correctly.

**Comment by michaelhhogue** [comment]
@ahsin-s From [README.md](https://github.com/OthersideAI/self-operating-computer/tree/main/README.md):

> **Note:** GPT-4V's error rate in estimating XY mouse click locations is currently quite high. This framework aims to track the progress of multimodal models over time, aspiring to achieve human-level performance in computer operation.

**Comment by ahsin-s** [comment]
There are demo videos showing this working with gpt4 and they seemed to at least get the model to click the address bar but I'm not sure if that was by placing the browser where the model expects it to be. Is the state of the art multimodal model still unable to achieve a simple address bar click?

**Comment by KBB99** [comment]
@ahsin-s this is a known limitation of GPT-4V at the moment. From [OpenAI Documentation,](https://platform.openai.com/docs/guides/vision/limitations) "*Limitations:* [...] Spatial reasoning: The model struggles with tasks requiring precise spatial localization, such as identifying chess positions."

**Comment by fpapleux** [comment]
On my machines, Linux, MacOs and Windows, I cannot get a single address bar click correct. I worked interactively with GPT4 uploading a screenshot and asking "give me the coordinates of the center of the 'New Mail' button in Outlook" and I received absolute coordinates that were pretty accurate. Would it help to shift the model to use absolute coordinates rather than %?

**Comment by bytecod3r** [comment]
> On my machines, Linux, MacOs and Windows, I cannot get a single address bar click correct. I worked interactively with GPT4 uploading a screenshot and asking "give me the coordinates of the center of the 'New Mail' button in Outlook" and I received absolute coordinates that were pretty accurate. Would it help to shift the model to use absolute coordinates rather than %?

It seems GPT4 can estimate the coordinates based on its own data and assumptions, and not using your screenshot.

![image](https://github.com/OthersideAI/self-operating-computer/assets/8350933/452e4dfa-6cfd-4cfd-a647-db63337bd3ef)

**Comment by ahsin-s** [comment]
That is the crux of the issue. Project maintainers don't seem to point out that gpt4v is NOT leveraging the grid overlay to estimate coordinates and is relying on heuristics in its training data instead. It does not use the image at all to infer where to place the mouse cursor.

**Comment by khalidovicGPT** [comment]
Hi, 

Until OpenAI improves the accuracy of pointer location, wouldn't it be wise to utilize keyboard shortcuts in Windows for easier navigation?

**Comment by joshbickett** [comment]
> @ahsin-s this is a known limitation of GPT-4V at the moment. From [OpenAI Documentation,](https://platform.openai.com/docs/guides/vision/limitations) "**Limitations:** [...] Spatial reasoning: The model struggles with tasks requiring precise spatial localization, such as identifying chess positions."

Hadn't seen this in their docs, thanks for pointing out

**Comment by joshbickett** [comment]
> Project maintainers don't seem to point out that gpt4v is NOT leveraging the grid overlay to estimate coordinates and is relying on heuristics in its training data instead.

GPT-4-V does in fact leverage the grid. This is the type of screenshot (below) that is sent to the API. When using the project a lot you'll notice GPT-4-V sometimes decides to use the exact cross sections of the grid to click (which is poor logic). 
![image](https://github.com/OthersideAI/self-operating-computer/assets/42594239/00a64f46-ce77-4398-9325-188c33f1a7ad)

Here's the section of the code where the grid is sent to GPT-4-v

```
def get_next_action_from_openai(messages, objective, accurate_mode):
...
    try:
        screenshots_dir = "screenshots"
        if not os.path.exists(screenshots_dir):
            os.makedirs(screenshots_dir)

        screenshot_filename = os.path.join(screenshots_dir, "screenshot.png")
        # Call the function to capture the screen with the cursor
        capture_screen_with_cursor(screenshot_filename)

        new_screenshot_filename = os.path.join(
            "screenshots", "screenshot_with_grid.png"
        )

        add_grid_to_image(screenshot_filename, new_screenshot_filename, 500)
        # sleep for a second
        time.sleep(1)

        with open(new_screenshot_filename, "rb") as img_file:
            img_base64 = base64.b64encode(img_file.read()).decode("utf-8")

        previous_action = get_last_assistant_message(messages)

        vision_prompt = format_vision_prompt(objective, previous_action)

        vision_message = {
            "role": "user",
            "content": [
                {"type": "text", "text": vision_prompt},
                {
                    "type": "image_url",
                    "image_url": {"url": f"data:image/jpeg;base64,{img_base64}"},
                },
            ],
        }

...

        response = client.chat.completions.create(
            model="gpt-4-vision-preview",
```

**Comment by Zero-coder** [comment]
> > Project maintainers don't seem to point out that gpt4v is NOT leveraging the grid overlay to estimate coordinates and is relying on heuristics in its training data instead.
> 
> GPT-4-V does in fact leverage the grid. This is the type of screenshot (below) that is sent to the API. When using the project a lot you'll notice GPT-4-V sometimes decides to use the exact cross sections of the grid to click (which is poor logic).
> 
> ![image](https://private-user-images.githubusercontent.com/42594239/292242796-00a64f46-ce77-4398-9325-188c33f1a7ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM0NDE1NjAsIm5iZiI6MTcwMzQ0MTI2MCwicGF0aCI6Ii80MjU5NDIzOS8yOTIyNDI3OTYtMDBhNjRmNDYtY2U3Ny00Mzk4LTkzMjUtMTg4YzMzZjFhN2FkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI0VDE4MDc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkNzY1MDRhOTc3ZDJjMWEzZmNiYmIwODZlYzA3ZjZmNDExOWI2NGY3MGQ5OTE1YjQ4MjFhOGI2NmIzMDE1ZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TYy6lJDPA35_gj8vz1EU_GN8MJbW8xsEqE_75PNWKKk)
> 
> Here's the section of the code where the grid is sent to GPT-4-v
> 
> ```
> def get_next_action_from_openai(messages, objective, accurate_mode):
> ...
>     try:
>         screenshots_dir = "screenshots"
>         if not os.path.exists(screenshots_dir):
>             os.makedirs(screenshots_dir)
> 
>         screenshot_filename = os.path.join(screenshots_dir, "screenshot.png")
>         # Call the function to capture the screen with the cursor
>         capture_screen_with_cursor(screenshot_filename)
> 
>         new_screenshot_filename = os.path.join(
>             "screenshots", "screenshot_with_grid.png"
>         )
> 
>         add_grid_to_image(screenshot_filename, new_screenshot_filename, 500)
>         # sleep for a second
>         time.sleep(1)
> 
>         with open(new_screenshot_filename, "rb") as img_file:
>             img_base64 = base64.b64encode(img_file.read()).decode("utf-8")
> 
>         previous_action = get_last_assistant_message(messages)
> 
>         vision_prompt = format_vision_prompt(objective, previous_action)
> 
>         vision_message = {
>             "role": "user",
>             "content": [
>                 {"type": "text", "text": vision_prompt},
>                 {
>                     "type": "image_url",
>                     "image_url": {"url": f"data:image/jpeg;base64,{img_base64}"},
>                 },
>             ],
>         }
> 
> ...
> 
>         response = client.chat.completions.create(
>             model="gpt-4-vision-preview",
> ```
> 
> Ultimately the Grid allows this project to be a proof of concept until spacial reasoning or object detention is better

‘Ultimately the Grid allows this project to be a proof of concept until spacial reasoning or object detention is better’ you mean object detection right?

**Comment by khalidovicGPT** [comment]
Hello @joshbickett,

I have some exciting news to share about the development of your self-operating-computer application. Apple has recently introduced an innovative AI model called Apple ML FERRET, which could be an ideal solution for the challenges you're facing with the GPT4-Vision model, especially regarding the localization of buttons in a screenshot.

The solution is open source but very GPU intensive: https://github.com/apple/ml-ferret


![localisation](https://github.com/OthersideAI/self-operating-computer/assets/131557521/a862bf38-4284-4cc2-9f8d-4611e024e160)

**Comment by joshbickett** [comment]
@khalidovicGPT yes, I saw that. It is certainly an exciting update. Would you be interested in attempting a PR of Ferret into the Self-Operating Computer?

**Comment by khalidovicGPT** [comment]
Hello @joshbickett, Happy New Year 2024! I am very honored to have the opportunity to contribute to the development of this remarkable application. The beginning of this year is quite busy for me in terms of availability, but I will do my best to help. I am going to inquire about the possibility of obtaining a server equipped with the ML FERRET model, with configured API access. I will keep you updated on my progress.

**Comment by joshbickett** [comment]
@khalidovicGPT sounds good. I briefly tried to run Ferret to try it out but realized it required more time that I had yesterday morning. From briefly reading the paper it sounds like it is accurate at detecting objects by X & Y pixel coordinates. It sounds like it could be compatible with the project. Interested to see if you can get it working! I wonder if anyone has an inference endpoint available for Ferret yet? We could use an established endpoint and have a key environmental variable

**Comment by khalidovicGPT** [comment]
Hi @joshbickett , good ! you have done more than me. I will update if I progress. Good luck!

**Comment by khalidovicGPT** [comment]
Hi Josh !

Hope you're doing well. Have you made any progress on the Ferret model?

On my end, I've been exploring a lighter and more efficient model: Vstar.

Source:
https://github.com/penghao-wu/vstar

You can test it here:
https://craigwu-vstar.hf.space/

The documentation:
https://vstar-seal.github.io/

There's even a YouTube video on this model if you understand Spanish:
https://youtu.be/4goxXl2Lk_M?si=gcthpZt8Y5DjxOkZ

I tried installing it locally, but it seems it doesn't work on PowerShell, only on a Linux OS with CUDA version lower than 2.

I'll let you see if you can make more progress than me.

I'll keep you updated if I manage to make more progress on my end. Good luck, Josh :)

Here are some tests that were conducted:

![puissance](https://github.com/OthersideAI/self-operating-computer/assets/131557521/5a8df73e-f2ab-4f80-8bdc-399e6eff2090)

![resultat2](https://github.com/OthersideAI/self-operating-computer/assets/131557521/11f9ff04-9913-49e7-92c6-28295c26f49b)

![classement](https://github.com/OthersideAI/self-operating-computer/assets/131557521/e78fc669-a00c-4821-ae9c-0fd25d543d72)

**Comment by joshbickett** [comment]
I wanted to mention the [OCR approach](https://x.com/josh_bickett/status/1750569783568011364?s=20) for those who have not seen it, it goes part way to solve this issue.

**Comment by joshbickett** [comment]
> I'll keep you updated if I manage to make more progress on my end. Good luck, Josh :

@khalidovicGPT curious if you were able to make any more progress?

**Comment by joshbickett** [comment]
I'll close this ticket for now since we have the OCR approach now!
